### DISCUSSION TITLE
Representing products

### DISCUSSION THREAD

**Discussion by bgamari** [clash-lang/clash-compiler#2646] [General]
One annoyance that I sometimes encounter is impedance mismatches due to the dual treatment of products in Clash. That is, a product can be represented as either a signal-of-a-product or as a product-of-signals (which are isomorphic as witnessed by `Bundle`). Under the status quo, product types will generally favor one of these representations, with the other being rather represented as a tuple. This asymmetry is awkward and compromises type safety. 

A pattern that I have used in my projects is to represent products as higher-kinded type. For instance, one might have a simple product:
```haskell
data Spi_M2S f
    = Spi_M2S { spiSclk :: f Bit
              , spiCs   :: f Bit
              , spiMosi :: f Bit
              }
```
where one can then provide a `Bundle` instance:
```haskell
instance Bundle (Spi_M2S Identity) where
    type Unbundled dom (Spi_M2S Identity) = Spi_M2S (Signal dom)
    bundle = ...
    unbundle = ...
```
I have been tempted to write a TH splice to derive such instances in the past.

Of course, this is a slightly unfortunate tradeoff as we must now manually wrap/unwrap `Identity`s  whenever working with the bundled type. I often find myself minimizing this syntactic overhead by locally binding `mkI = Identity` and `unI = runIdentity`. 
However, another design would be to provide pattern synonyms to handle the common bundled case:
```haskell
data Spi_M2S_F f
    = Spi_M2S_F { spiSclk :: f Bit
              , spiCs   :: f Bit
              , spiMosi :: f Bit
              }

type Spi_M2S = Spi_M2S_F Identity

pattern Spi_M2S :: Bit -> Bit -> Bit -> Spi_M2S
pattern Spi_M2S a b c = Spi_M2S_F { spiSclk = Identity a, spiCs = Identity b, spiMosi=Identity c }
```
This is another tricky set of trade-offs as now we rely on the `DuplicateRecordFields` and `PatternSynonyms` extensions. Is this worth it? I don't know.

Have others thought about this problem? Is there a nicer solution that I have missed?

**Comment by gergoerdi** [comment]
I've been using Barbies for this, so that the unbundled type can have no wrappers on the field types:

https://github.com/gergoerdi/retroclash-lib/blob/master/src%2FRetroClash%2FBarbies.hs

**Reply by bgamari** [reply]
That is a very useful reference. Thanks!

**Reply by gergoerdi** [reply]
> That is a very useful reference. Thanks!

You might also like this usage example, then: https://github.com/gergoerdi/retroclash-lib/blob/master/src%2FRetroClash%2FCPU.hs
